File: filemaid/cleaners/pip.py
"""pip cache cleanup."""
import subprocess
from pathlib import Path

from filemaid.cleaners._size import dir_size, humanize


def _pip_cache_dir() -> Path:
    try:
        result = subprocess.run(
            ["python3", "-m", "pip", "cache", "dir"],
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            text=True,
            timeout=30,
        )
        path = result.stdout.strip()
        if path:
            return Path(path).expanduser()
    except Exception:
        pass
    return Path.home() / "Library" / "Caches" / "pip"


def can_run() -> bool:
    return True


def run(dry_run: bool, config: dict) -> str:
    cache_dir = _pip_cache_dir()
    before = dir_size(cache_dir) if cache_dir.exists() else 0
    if dry_run:
        return f"pip: would purge cache (~{humanize(before)} to remove)"
    try:
        result = subprocess.run(
            ["python3", "-m", "pip", "cache", "purge"],
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            timeout=120,
        )
        after = dir_size(cache_dir) if cache_dir.exists() else 0
        saved = max(0, before - after)
        output = result.stdout.strip()
        return f"pip: cache purged, saved {humanize(saved)}\n{output}"
    except Exception as exc:
        return f"pip: failed - {exc}"

File: internal/cleaners/pip.go
package cleaners

import (
	"path/filepath"
	"strings"

	"github.com/logicminds/filemaid/internal/config"
)

var pipCleaner = Cleaner{
	Name:   "pip",
	CanRun: pipCanRun,
	Run:    pipRun,
}

func pipCanRun() bool {
	return true
}

func pipCacheDir() string {
	out, err := runner.Run("python3", "-m", "pip", "cache", "dir")
	if err == nil {
		path := strings.TrimSpace(string(out))
		if path != "" {
			return expandPath(path)
		}
	}
	return filepath.Join(home(), "Library", "Caches", "pip")
}

func pipRun(dryRun bool, cfg *config.Config) CleanupResult {
	cacheDir := pipCacheDir()
	before := dirSizeIfExists(cacheDir)

	if dryRun {
		saved := before
		return CleanupResult{
			Name:       "pip",
			Status:     "dry-run",
			Saved:      &saved,
			SavedHuman: Humanize(before),
			Detail:     "would purge cache",
		}
	}

	out, err := runner.Run("python3", "-m", "pip", "cache", "purge")
	if err != nil {
		return CleanupResult{
			Name:   "pip",
			Status: "failed",
			Detail: err.Error(),
		}
	}

	after := dirSizeIfExists(cacheDir)
	saved := before - after
	if saved < 0 {
		saved = 0
	}
	output := strings.TrimSpace(string(out))
	detail := "cache purged"
	if output != "" {
		detail = "cache purged\n" + output
	}

	return CleanupResult{
		Name:       "pip",
		Status:     "ok",
		Saved:      &saved,
		SavedHuman: Humanize(saved),
		Detail:     detail,
	}
}
